### PR TITLE
Add Google Cloud Pub/Sub Spark Streaming Connector

### PR DESCRIPTION
**Pull request by pmkc**

### CODE REVIEW

**Comment by kir-titievsky** [comment]
@adithyagirish It seems we need your approval here.

**Comment by pmkc** [comment]
Adithya (@adithyagirish) is currently travelling without access to github, but has confirmed via email that he consents to his code being contributed to the repo.

File: pubsub/src/main/scala/com/google/cloud/spark/pubsub/PubsubReceiver.scala [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+          throw failure
+        }
+      },
+      MoreExecutors.directExecutor()
+    )
+  }
+
+  def onStart() {
+    init()
+    subscriber.startAsync.awaitRunning()
```
Comment (AngusDavis) [inline pubsub/src/main/scala/com/google/cloud/spark/pubsub/PubsubReceiver.scala]: Per docs, onStart, onStop must not block indefinitely.

File: pubsub/src/main/scala/com/google/cloud/spark/pubsub/PubsubReceiver.scala [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+    .build
+
+    // Until stopped or connection broken continue reading
+    val receiver = new PubsubMessageReceiver(this)
+    subscriber = buildSubscriber(subscription, receiver)
+
+    subscriber.addListener(
+      new ApiService.Listener {
+        override def failed(from: State, failure: Throwable) {
+          log.error(failure.getMessage)
```
Comment (AngusDavis) [inline pubsub/src/main/scala/com/google/cloud/spark/pubsub/PubsubReceiver.scala]: Per the docs, reportError should be called to inform the driver of this error.

File: pubsub/src/main/scala/com/google/cloud/spark/pubsub/SparkPubsubWriters.scala [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+
+/**
+ * Allows for writing RDD records to Pub/Sub
+ *
+ * To complete the pipeline, this will allows
+ * users to take records they've processed in Spark
+ * and write them back to a Pub/Sub topic
+ */
+class DStreamPubsubWriter(dStream: DStream[String]) extends Serializable {
+  def writeToPubsub(clientProvider: () => PubsubClient): Unit = {
```
Comment (AngusDavis) [inline pubsub/src/main/scala/com/google/cloud/spark/pubsub/SparkPubsubWriters.scala]: + docs.

File: pubsub/src/main/scala/com/google/cloud/spark/pubsub/stackdriver/PubsubMetrics.scala [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+ * num_undelivered_messages
+ */
+class PubsubMetrics(
+  projectId: String,
+  metricServiceClientProvider: () => MetricServiceClient = MetricServiceClient.create,
+  listTimeSeries:
+    (MetricServiceClient, ListTimeSeriesRequest) => Iterable[TimeSeries] =
+    PubsubMetrics.listTimesSeries) {
+
+  val name = ProjectName.create(projectId)
```
Comment (AngusDavis) [inline pubsub/src/main/scala/com/google/cloud/spark/pubsub/stackdriver/PubsubMetrics.scala]: private val

File: pubsub/src/main/scala/com/google/cloud/spark/pubsub/PubsubClient.scala [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+    /*
+    Using same structure as recommended by Google PubSub Java Docs
+    https://cloud.google.com/pubsub/docs/publisher#pubsub-publish-java
+     */
+    val topic = TopicName.create(projectId, topicId)
+    var publisher: Publisher = null
+    val messageIdFutures: ListBuffer[ApiFuture[String]] = ListBuffer.empty[ApiFuture[String]]
+
+    try {
+      publisher = Publisher.defaultBuilder(topic).build
```
Comment (AngusDavis) [inline pubsub/src/main/scala/com/google/cloud/spark/pubsub/PubsubClient.scala]: Publishers are really expensive to create (grpc setup). You will most likely see really bad perf on this method.

**Comment by pmkc** [comment]
I will follow up with suggested changes in subsequent PR. Merging as is for now.

**Pull request merged**